In [2]:
import io
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from google.colab import files

# ==========================================
# 1. FILE UPLOAD PROMPT
# ==========================================
print("Click the button below to upload your web traffic dataset.")
uploaded = files.upload()

# Get the filename and load the data
filename = list(uploaded.keys())[0]
df = pd.read_csv(io.BytesIO(uploaded[filename]))

# Ensure the column names match perfectly (stripping spaces)
df.columns = df.columns.str.strip()

print(f"\nSuccessfully loaded '{filename}'!")
print(f"Detected Columns: {list(df.columns)}\n")

# ==========================================
# 2. TRAFFIC OVERVIEW & ENGAGEMENT
# ==========================================
print("=== [1] CORE WEB TRAFFIC METRICS ===")
# Calculate overall averages from your metrics
total_views = df['Page Views'].sum() if 'Page Views' in df.columns else "N/A"
avg_duration = df['Session Duration'].mean() if 'Session Duration' in df.columns else "N/A"
avg_bounce = df['Bounce Rate'].mean() if 'Bounce Rate' in df.columns else "N/A"
avg_conversion = df['Conversion Rate'].mean() if 'Conversion Rate' in df.columns else "N/A"

print(f"🔹 Total Page Views:          {total_views:,}")
print(f"🔹 Average Session Duration:  {avg_duration:.2f} minutes/seconds")
print(f"🔹 Average Bounce Rate:      {avg_bounce:.2%}" if isinstance(avg_bounce, float) and avg_bounce <= 1 else f"🔹 Average Bounce Rate:      {avg_bounce:.2f}%")
print(f"🔹 Average Conversion Rate:  {avg_conversion:.2%}" if isinstance(avg_conversion, float) and avg_conversion <= 1 else f"🔹 Average Conversion Rate:  {avg_conversion:.2f}%")
print("\n" + "="*40 + "\n")

# ==========================================
# 3. TRAFFIC SOURCE ANALYSIS
# ==========================================
if 'Traffic Source' in df.columns:
    print("=== [2] TRAFFIC PERFORMANCE BY SOURCE ===")
    # Grouping performance metrics by how users arrived
    source_summary = df.groupby('Traffic Source').agg({
        'Page Views': 'sum',
        'Session Duration': 'mean',
        'Conversion Rate': 'mean'
    }).sort_values(by='Page Views', ascending=False)

    display(source_summary)
    print("\n" + "="*40 + "\n")

# ==========================================
# 4. IDENTIFYING DROP-OFF & RETENTION
# ==========================================
print("=== [3] HIGH DROP-OFF / BOUNCE ANALYSIS ===")
if 'Bounce Rate' in df.columns:
    # Sort to find pages or sources with the highest bounce/drop-off rate
    high_dropoff = df.sort_values(by='Bounce Rate', ascending=False).head(5)
    print("Top 5 Areas with the Highest Drop-off (Bounce Rate):")
    display(high_dropoff)
else:
    print("Bounce Rate column missing or improperly formatted.")
print("\n" + "="*40 + "\n")

# ==========================================
# 5. INTERACTIVE PERFORMANCE VISUALIZATION
# ==========================================
print("=== [4] TRAFFIC VS CONVERSION VISUALIZATION ===")
# Creates a scatter map analyzing traffic volume vs how well it converts

# Check if we can map by Traffic Source or look at overall rows
color_param = 'Traffic Source' if 'Traffic Source' in df.columns else None

fig = px.scatter(
    df,
    x='Page Views',
    y='Conversion Rate',
    size='Session Duration' if 'Session Duration' in df.columns else None,
    color=color_param,
    hover_data=['Bounce Rate', 'Time on Page'] if 'Time on Page' in df.columns else ['Bounce Rate'],
    title="Traffic Volume vs. Conversion Rate (Size = Session Duration)",
    template="plotly_dark"
)

fig.update_layout(height=500)
fig.show()

Click the button below to upload your web traffic dataset.


Saving website_wata.csv to website_wata (1).csv

Successfully loaded 'website_wata (1).csv'!
Detected Columns: ['Page Views', 'Session Duration', 'Bounce Rate', 'Traffic Source', 'Time on Page', 'Previous Visits', 'Conversion Rate']

=== [1] CORE WEB TRAFFIC METRICS ===
🔹 Total Page Views:          9,901
🔹 Average Session Duration:  3.02 minutes/seconds
🔹 Average Bounce Rate:      28.48%
🔹 Average Conversion Rate:  98.21%


=== [2] TRAFFIC PERFORMANCE BY SOURCE ===


,Page Views,Session Duration,Conversion Rate
Traffic Source,,,
Organic,3950,3.104197,0.982282
Paid,2116,2.942816,0.979009
Referral,1500,3.125496,0.987727
Social,1263,3.057273,0.982679
Direct,1072,2.692060,0.978679




=== [3] HIGH DROP-OFF / BOUNCE ANALYSIS ===
Top 5 Areas with the Highest Drop-off (Bounce Rate):


,Page Views,Session Duration,Bounce Rate,Traffic Source,Time on Page,Previous Visits,Conversion Rate
50,5,0.423207,0.844939,Organic,0.727810,4,0.749204
1243,5,0.357272,0.837279,Social,9.323701,1,1.000000
1311,4,0.631288,0.835646,Organic,4.169423,2,1.000000
1265,5,1.510904,0.802257,Social,2.608188,2,1.000000
253,3,2.720047,0.792655,Organic,2.402637,4,1.000000




=== [4] TRAFFIC VS CONVERSION VISUALIZATION ===


Web Traffic Analytics Report :

1.Executive Summary:

  An analysis of 9,901 total page views reveals a highly sticky web ecosystem characterized by an exceptional average baseline conversion rate ($98.21\%$) and a healthy, controlled average bounce rate ($28.48\%$). Visitors maintain a steady average session duration of 3.02 units, indicating high intent and prompt engagement.While core performance across all baseline channels remains robust, granular inspection identifies isolated friction points within Organic and Social traffic segments, where localized drop-off anomalies spike as high as $84.49\%$.

2.Core Performance Indicators (KPIs):

The platform demonstrates premium engagement across all core metrics:

    Metric                      System Value      Performance Status

    Volume (Total Page Views)   9,901             Strong baseline visibility

    
    Engagement                  3.02 units        Stable, sustained user
    (Avg. Session Duration)                       interest

    Retention                   28.48%            Excellent (well below industry
    (Avg. Bounce Rate)                            risk thresholds of $>50\%$)
                                                     

    Success                     98.21%            Hyper-optimized conversion
    (Avg. Conversion Rate)                        tracking
3.Acquisition Channel Analysis:

Traffic acquisition is well-diversified, led heavily by inbound discovery:

Organic Traffic is your primary volume driver (3,950 page views) and retains users effectively with a strong session duration of 3.10.

Referral Traffic represents your highest-quality audience, yielding the peak conversion rate of $98.77\%$ and the longest time spent on-site (3.12).

Direct Traffic presents the lowest depth of engagement, trailing with a lower session duration of 2.69 and a lower relative conversion efficiency ($97.87\%$).

Channel Breakdown Table:

    Traffic         Volume                   Average             Average
    Source          (Page Views)          Session Duration     Conversion Rate
      
    
    Organic         3,950                     3.10               98.23%
    Paid            2,116                     2.94               97.90%
    Referral        1,500                     3.13               98.77%
    Social          1,263                     3.06               98.27%
    Direct          1,072                     2.69               97.87%

4.Leakage & Drop-Off Mapping:

  Despite optimal sitewide averages, a deep-dive audit into the Top 5 highest bounce configurations isolates critical experience friction:

  ⚠️ Critical Alert: Specific landing segments are experiencing severe bounce rates ranging from $79.27\%$ to $84.49\%$.

  High-Risk Friction Clusters:

  1.The Organic Search Mismatch: The highest bounce rate recorded ($84.49\%$) occurs within an Organic segment where users spend less than 1 unit of time on-site, despite having previous history (4 visits). This points to an index mismatch or broken page state for that specific entry query.

  2:Social Media Quick-Exits: Social traffic captures a segment with an extreme bounce rate ($83.73\%$), where users immediately leave despite registering a perfect conversion metric elsewhere, indicating a highly volatile drop-off behavior.

5.Strategic Recommendations:

Audit High-Bounce Organic Content: Investigate the specific target landing environments mapped to Organic sources showing an $84.49\%$ drop-off. Audit for page load speeds, broken visual components, or mismatched keyword intent.

Optimize Direct Traffic Landing Experience: Because Direct traffic demonstrates lower baseline sessions (2.69), simplify navigation pathways or deploy immediate interactive cues on the home/landing destinations to capture returning or brand-aware users instantly.

Replicate Referral Strategies: Referral loops are generating your highest structural returns ($98.77\%$ conversion). Identify top backlink sources and scale collaborative content partnerships with those domains.